### Ejercicios de la S4

#### Reto 1: Series

##### 1. Objetivos:
    - Practicar la creación de `Series` y la indexación básica de éstas
 
---
    
##### 2. Desarrollo:

In [1]:
## Realiza aquí los imports que necesites para que tu código funcione correctamente
import pandas as pd

##### a) Creación de `Series`

A continuación tenemos unas variables que contienen los nombres de los ejecutivos más importantes de nuestra ya conocida EyePoker Inc. Debajo de eso hay una variable que no ha sido asignada aún:

In [2]:
ejecutivo_1 = 'Marco P.'
ejecutivo_2 = 'Jenny'
ejecutivo_3 = 'Britney Baby'
ejecutivo_4 = 'Pepe Guardabosques'
ejecutivo_5 = 'Lombardo El Destructor'

ejecutivos = pd.Series([ejecutivo_1, ejecutivo_2, ejecutivo_3, ejecutivo_4, ejecutivo_5])

sueldos = pd.Series([7500, 8400, 9200, 7300, 8000], index = ejecutivos)

Tu tarea es crear una `Serie` de `pandas` y asignarla a la variable `sueldos`. La información que hay dentro de esta variable son (oh, sorpresa) los sueldos de dichos ejecutivos.

Los valores de la `Serie` serán los sueldos, mientras que el índice de la `Serie` serán los nombres de los ejecutivos.

Crea una `Serie` y asígnala a `sueldos` de manera que el código que tenemos debajo funcione correctamente:

In [3]:
print('== Sueldos de los principales ejecutivos de EyePoker Inc. ==\n')

print(f'{("Ejecutivo"):25} | {("Sueldo")}')
print('----------------------------------------')
print(f'{ejecutivo_1:25} | ${(sueldos.loc[ejecutivo_1])} MXN')
print(f'{ejecutivo_2:25} | ${(sueldos.loc[ejecutivo_2])} MXN')
print(f'{ejecutivo_3:25} | ${(sueldos.loc[ejecutivo_3])} MXN')
print(f'{ejecutivo_4:25} | ${(sueldos.loc[ejecutivo_4])} MXN')
print(f'{ejecutivo_5:25} | ${(sueldos.loc[ejecutivo_5])} MXN')

== Sueldos de los principales ejecutivos de EyePoker Inc. ==

Ejecutivo                 | Sueldo
----------------------------------------
Marco P.                  | $7500 MXN
Jenny                     | $8400 MXN
Britney Baby              | $9200 MXN
Pepe Guardabosques        | $7300 MXN
Lombardo El Destructor    | $8000 MXN


#### Reto 2: Indexación de Series

##### 1. Objetivos:
    - Practicar técnicas avanzadas de indexación de `Series`
 
---
    
##### 2. Desarrollo:

##### a) Indexación de `Series`

Tenemos una `Serie` que contiene los gastos mensuales totales (en MXN) de distintas divisiones de EyePoker Inc. Tú eres el Contador Oficial y tienes que obtener subconjuntos de datos que sirvan para agregar los gastos totales de diferentes combinaciones de divisiones.

Los datos son los siguientes:

In [4]:
import pandas as pd

gastos_mensuales = {
    'A': 15000,
    'B': 200000,
    'C': 3250000,
    'D': 120000,
    'E': 135000,
    'F': 55000,
    'G': 100000,
    'H': 25000
}

gastos_serie = pd.Series(gastos_mensuales)

In [5]:
gastos_serie

A      15000
B     200000
C    3250000
D     120000
E     135000
F      55000
G     100000
H      25000
dtype: int64

El índice es el nombre de la división y los valores son los gastos mensuales en MXN.

Indexando la serie gastos_serie extrae las combinaciones de divisiones que se indican debajo para poder hacer los cálculos necesarios. El primero es un ejemplo:

In [6]:
# Los gastos de la división 'D' y 'G'
gastos_D_G = gastos_serie[['D', 'G']]

# Los gastos de la división 'A' y 'E'
gastos_A_E = gastos_serie[['A', 'E']]

# Los gastos de la división 'B', 'F' y 'H'
gastos_B_F_H = gastos_serie[['B', 'F', 'H']]

# Los gastos desde la primera división hasta la división 'E'
gastos_principio_a_E = gastos_serie[:'E']

# Los gastos desde la división 'D' hasta la 'G'
gastos_D_a_G = gastos_serie['D':'G']

# Los gastos desde la división 'C' hasta el la última división
gastos_C_a_final = gastos_serie['C':]

Corre la función de verificación revisar_indexaciones para verificar tu resultado:

In [7]:
# Pega aquí la función de verificación
def revisar_indexaciones(gastos_serie, gastos_D_G, gastos_A_E, gastos_B_F_H,
                         gastos_principio_a_E, gastos_D_a_G, gastos_C_a_final):
    
    print(f'== Revisión de Indexaciones ==\n')
    print(f"{'Indexación':30} | {'Resultado':15} | {'Suma esperada ':15} | {'Suma recibida ':15}")
    print("-"*85)
    revisar_indexacion(gastos_serie.loc[['D', 'G']], gastos_D_G, 'División D y G')
    revisar_indexacion(gastos_serie.loc[['A', 'E']], gastos_A_E, 'División A y E')
    revisar_indexacion(gastos_serie.loc[['B', 'F', 'H']], gastos_B_F_H, 'División B, F y H')
    revisar_indexacion(gastos_serie.loc[:'E'], gastos_principio_a_E, 'Desde primera División a E')
    revisar_indexacion(gastos_serie.loc['D':'G'], gastos_D_a_G, 'División D y G')
    revisar_indexacion(gastos_serie.loc['C':], gastos_C_a_final, 'División C a última División')

def formatear_precio(precio):
    return f"${precio} MXN"
    
def revisar_indexacion(esperada, recibida, nombre):
    es_correcta = 'Correcta' if esperada.equals(recibida) else 'Incorrecta'
    suma_esperada = formatear_precio(sum(esperada))
    suma_recibida = formatear_precio(sum(recibida))
    print(f"{nombre:30} | {es_correcta:15} | {suma_esperada:15} | {suma_recibida:15}")

revisar_indexaciones(gastos_serie, gastos_D_G, gastos_A_E, gastos_B_F_H,
                         gastos_principio_a_E, gastos_D_a_G, gastos_C_a_final)

== Revisión de Indexaciones ==

Indexación                     | Resultado       | Suma esperada   | Suma recibida  
-------------------------------------------------------------------------------------
División D y G                 | Correcta        | $220000 MXN     | $220000 MXN    
División A y E                 | Correcta        | $150000 MXN     | $150000 MXN    
División B, F y H              | Correcta        | $280000 MXN     | $280000 MXN    
Desde primera División a E     | Correcta        | $3720000 MXN    | $3720000 MXN   
División D y G                 | Correcta        | $410000 MXN     | $410000 MXN    
División C a última División   | Correcta        | $3685000 MXN    | $3685000 MXN   


#### Reto 3: DataFrames

##### 1. Objetivos:
    - Aprender a crear `DataFrames` e indexar por columna y por fila
 
---
    
##### 2. Desarrollo:

##### a) Creación e indexación de `DataFrames`

Eres el Data Wrangler (procesador de datos) de EyePoker Inc. Tienes el siguiente diccionario con datos que se refieren a diferentes productos que vende la empresa. Este es tu conjunto de datos y el índice que le corresponde:

In [8]:
datos_productos = {
    "nombre": ["Pokemaster", "Cegatron", "Pikame Mucho", "Lazarillo de Tormes", "Stevie Wonder", "Needle", "El AyMeDuele"],
    "precio": [10000, 5500, 3500, 750, 15500, 12250, 23000],
    "peso": [1.2, 1.5, 2.3, 5.5, 3.4, 2.4, 8.8],
    "capacidad de destrucción retinal": [3, 7, 6, 8, 9, 2, 10],
    "disponible": [True, False, True, True, False, False, True]
}

indice = [1, 2, 3, 4, 5, 6, 7]

El Analista de Datos de la empresa quiere realizar algunas visualizaciones con este conjunto de datos, pero no es muy buen procesador de datos, así que te pide a ti ayuda para crear los subconjuntos de datos que necesita para sus visualizaciones.

Te ha dado las descripciones de los subconjuntos que necesita, en el orden en el que los quiere.

Tu primer paso es convertir tu `diccionario` a `DataFrame` usando `datos_productos` e `indice`:

In [9]:
df_productos = pd.DataFrame(datos_productos, index = indice)

In [10]:
df_productos

nombre  precio  peso  capacidad de destrucción retinal  \
1           Pokemaster   10000   1.2                                 3   
2             Cegatron    5500   1.5                                 7   
3         Pikame Mucho    3500   2.3                                 6   
4  Lazarillo de Tormes     750   5.5                                 8   
5        Stevie Wonder   15500   3.4                                 9   
6               Needle   12250   2.4                                 2   
7         El AyMeDuele   23000   8.8                                10   

   disponible  
1        True  
2       False  
3        True  
4        True  
5       False  
6       False  
7        True

Ahora, indexa tu DataFrame para obtener los subconjuntos requeridos. Los productos en existencia tienen un orden específico en la base de datos. El orden correcto es el que está definido en datos_productos. Eso significa que el "Pokemaster" tiene el índice 1 y es el primer producto; y el "El AyMeDuele" tiene el ìndice 7 y es el último producto.

Realiza las indexaciones debajo. Recuerda ordenar tus DataFrames en el orden en el que los menciona el Analista:

In [14]:
# Quiero un DataFrame que contenga los productos "Pikame Mucho" y "Stevie Wonder"
pm_sw = df_productos.loc[[3, 5]]

# Quiero un DataFrame que contenga desde el producto #4 hasta el último
p4_final = df_productos.loc[4:]

# Quiero un DataFrame que contenga los productos "El AyMeDuele", "Lazarillo de Tormes" y "Needle"
amd_lt_n = df_productos.loc[[7, 4, 6]]

# Quiero un DataFrame que contenga desde el primer producto hasta el producto #5
primer_p5 = df_productos.loc[:5]

# Quiero un DataFrame que contenga los productos "Pikame Mucho" y "Lazarillo de Tormes", pero sólo con las columnas "nombre", "precio" y "peso"
pm_lt_pp = df_productos.loc[[3, 4], ['nombre', 'precio', 'peso']]

# Quiero un DataFrame que contenga todos los productos pero con sólo las columnas 'nombre', 'precio' y 'capacidad de destrucción retinal'
t_pcdr = df_productos[['nombre', 'precio', 'capacidad de destrucción retinal']]

# Quiero un DataFrame que contenga desde el producto #3 hasta el #6, pero sólo las columnas 'nombre', 'precio' y 'disponible'
p3_p6_pd = df_productos[['nombre', 'precio', 'disponible']].loc[3:6]

Corre la función de verificación verificar_indexaciones para verificar tu resultado:

In [15]:
# Pega aquí la función de verificación
def verificar_indexaciones(datos_productos, indice, df_productos, pm_sw, p4_final, amd_lt_n, primer_p5, pm_lt_pp, t_pcdr, p3_p6_pd):
    
    import pandas as pd
    
    df_productos_esperado = pd.DataFrame(datos_productos, index=indice)
    if not df_productos_esperado.equals(df_productos):
        print(f'df_productos ha sido creado incorrectamente ... Favor de revisar')
        return
    
    print(f'== Verificación de Indexaciones ==\n')
    verificar_indexacion(df_productos.loc[[3, 5]], pm_sw, 'DataFrame que contenga los productos "Pikame Mucho" y "Stevie Wonder"')
    verificar_indexacion(df_productos.loc[4:], p4_final, 'DataFrame que contenga desde el producto #4 hasta el último')
    verificar_indexacion(df_productos.loc[[7, 4, 6]], amd_lt_n, 'DataFrame que contenga los productos "El AyMeDuele", "Lazarillo de Tormes" y "Needle"')
    verificar_indexacion(df_productos.loc[:5], primer_p5, 'DataFrame que contenga desde el primer producto hasta el producto #5')
    verificar_indexacion(df_productos.loc[[3, 4], ['nombre', 'precio', 'peso']], pm_lt_pp, 'DataFrame que contenga los productos "Pikame Mucho" y "Lazarillo de Tormes", pero sólo con las columnas "nombre", "precio" y "peso"')
    verificar_indexacion(df_productos[['nombre', 'precio', 'capacidad de destrucción retinal']], t_pcdr, "DataFrame que contenga todos los productos pero con sólo las columnas 'nombre', 'precio' y 'capacidad de destrucción retinal'")
    verificar_indexacion(df_productos.loc[3:6, ['nombre', 'precio', 'disponible']], p3_p6_pd, "DataFrame que contenga desde el producto #3 hasta el #6, pero sólo las columnas 'nombre', 'precio' y 'disponible'")
    
def verificar_indexacion(esperada, recibida, descripcion):
    es_correcta = "Correcto" if esperada.equals(recibida) else "Incorrecto"
    respuesta = "Muchas gracias!" if es_correcta == "Correcto" else "Favor de revisar"
    print(f"\n- Descripción de pedido: {descripcion}")
    print(f"El pedido es {es_correcta} ... {respuesta}")

verificar_indexaciones(datos_productos, indice, df_productos, pm_sw, p4_final, amd_lt_n, primer_p5, pm_lt_pp, t_pcdr, p3_p6_pd)

== Verificación de Indexaciones ==


- Descripción de pedido: DataFrame que contenga los productos "Pikame Mucho" y "Stevie Wonder"
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: DataFrame que contenga desde el producto #4 hasta el último
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: DataFrame que contenga los productos "El AyMeDuele", "Lazarillo de Tormes" y "Needle"
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: DataFrame que contenga desde el primer producto hasta el producto #5
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: DataFrame que contenga los productos "Pikame Mucho" y "Lazarillo de Tormes", pero sólo con las columnas "nombre", "precio" y "peso"
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: DataFrame que contenga todos los productos pero con sólo las columnas 'nombre', 'precio' y 'capacidad de destrucción retinal'
El pedido es Correcto ... Muchas gracias!

- Descripció

#### Reto 4: Manipulación de Columnas

##### 1. Objetivos:
    - Practicar asignar, reasignar y eliminar columnas de `DataFrames`
 
---
    
##### 2. Desarrollo:

##### a) Transformación de `DataFrames`

Eres de nuevo el Data Wrangler Inc. Tienes el mismo conjunto de datos que en el Reto pasado:

In [ ]:
df_productos

Esta vez el Analista de Datos te hace 3 nuevos pedidos que incluyen la creación de una nueva columna, la asignación de nuevos datos a una columna y la eliminación de un par de columnas. Crea un `DataFrame` usando `datos_productos` e `indice`, realiza sus pedidos y envíalos para su verificación:

In [16]:
# Agrega por favor una nueva columna a `df_productos_mas_columna_nueva` con el nombre de columna "nivel de dolor"
columna_nueva = [4, 7, 6, 8, 9, 7, 3]
df_productos_mas_columna_nueva = df_productos.copy()
df_productos_mas_columna_nueva['nivel de dolor'] = columna_nueva

# Cambia por favor el `DataFrame` `df_productos_descuento` cambiando la columna `precio` por la información contenida en `precios_descuento`
precios_descuento = [8000, 4000, 2000, 500, 14000, 10000, 15000]
df_productos_descuento = df_productos.copy()
df_productos_descuento['precio'] = precios_descuento

# Elimina por favor las columnas "precio" y "peso" de `df_productos` y asigna el resultado a `df_productos_sin_precio_ni_peso`
df_productos_sin_precio_ni_peso = df_productos.drop(columns = ['precio', 'peso'])

Corre la función de verificación verificar_modificaciones para verificar tu resultado:

In [17]:
# Pega aquí la función de verificación
def verificar_modificaciones(datos_productos, indice, df_productos, columna_nueva, df_productos_mas_columna_nueva, 
                             precios_descuento, df_productos_descuento, df_productos_sin_precio_ni_peso):
    
    import pandas as pd
    
    df_productos_esperado = pd.DataFrame(datos_productos, index=indice)
    if not df_productos_esperado.equals(df_productos):
        print(f'df_productos ha sido creado incorrectamente ... Favor de revisar')
        return
    
    print(f'== Verificación de Modificaciones ==\n')
    columna_nueva_serie = pd.Series(columna_nueva, index=indice)
    df_productos_mas_columna_nueva_2 = df_productos.copy()
    df_productos_mas_columna_nueva_2['nivel de dolor'] = columna_nueva_serie
    verificar_modificacion(df_productos_mas_columna_nueva_2, df_productos_mas_columna_nueva, 'Agrega por favor columna `columna_nueva` a `df_productos_mas_columna_nueva` con el nombre de columna "nivel de dolor"')
    
    precios_descuento_serie = pd.Series(precios_descuento, index=indice)
    df_productos_descuento_2 = df_productos.copy()
    df_productos_descuento_2['precio'] = precios_descuento_serie
    verificar_modificacion(df_productos_descuento_2, df_productos_descuento, 'Cambia por favor el `DataFrame` `df_productos_descuento` cambiando la columna `precio` por la información contenida en `precios_descuento`')
    
    df_productos_sin_precio_ni_peso_2 = df_productos.drop(columns=['precio', 'peso'])
    verificar_modificacion(df_productos_sin_precio_ni_peso_2, df_productos_sin_precio_ni_peso, 'Elimina por favor las columnas "precio" y "peso"')
    
def verificar_modificacion(esperada, recibida, descripcion):
    es_correcta = "Correcto" if esperada.equals(recibida) else "Incorrecto"
    respuesta = "Muchas gracias!" if es_correcta == "Correcto" else "Favor de revisar"
    print(f"\n- Descripción de pedido: {descripcion}")
    print(f"El pedido es {es_correcta} ... {respuesta}")

verificar_modificaciones(datos_productos, indice, df_productos, columna_nueva, df_productos_mas_columna_nueva, 
                             precios_descuento, df_productos_descuento, df_productos_sin_precio_ni_peso)

== Verificación de Modificaciones ==


- Descripción de pedido: Agrega por favor columna `columna_nueva` a `df_productos_mas_columna_nueva` con el nombre de columna "nivel de dolor"
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: Cambia por favor el `DataFrame` `df_productos_descuento` cambiando la columna `precio` por la información contenida en `precios_descuento`
El pedido es Correcto ... Muchas gracias!

- Descripción de pedido: Elimina por favor las columnas "precio" y "peso"
El pedido es Correcto ... Muchas gracias!


#### Reto 5: Análisis Exploratorio de Datos

##### 1. Objetivos:
    - Practicar leer archivos JSON usando pandas
    - Practicar hacerse preguntas acerca de los conjuntos de datos que tenemos
 
---
    
##### 2. Desarrollo:

Vamos a practicar explorar conjuntos de datos y hacernos preguntas acerca de ellos.

Tenemos un conjunto de datos en formato JSON almacenado en '../Datasets/new_york_times_bestsellers-clean.json' (en la carpeta /Datasets en el directorio raíz del módulo).

Primero que nada, lee el archivo JSON y crea un `DataFrame` con él:

In [18]:
## Realiza tus imports
import json

In [19]:
## Lee aquí tu archivo JSON
f = open('../Datasets/new_york_times_bestsellers-clean.json', 'r')
data_json = json.load(f)
f.close()

In [20]:
## Crea aquí tu DataFrame
df = pd.DataFrame.from_dict(data_json)

In [21]:
df

amazon_product_url  \
0     http://www.amazon.com/The-Host-Novel-Stephenie...   
1     http://www.amazon.com/Love-Youre-With-Emily-Gi...   
2     http://www.amazon.com/The-Front-Garano-Patrici...   
3     http://www.amazon.com/Snuff-Chuck-Palahniuk/dp...   
4     http://www.amazon.com/Sundays-at-Tiffanys-Jame...   
...                                                 ...   
3028  http://www.amazon.com/Six-Years-Harlan-Coben/d...   
3029  http://www.amazon.com/The-Interestings-Novel-M...   
3030  http://www.amazon.com/Man-Without-Breath-Berni...   
3031  http://www.amazon.com/The-Storyteller-Jodi-Pic...   
3032  http://www.amazon.com/Z-A-Novel-Zelda-Fitzgera...   

                                        author  \
0                              Stephenie Meyer   
1                                 Emily Giffin   
2                            Patricia Cornwell   
3                              Chuck Palahniuk   
4     James Patterson and Gabrielle Charbonnet   
...                                        ...   
3028                              Harlan Coben   
3029                              Meg Wolitzer   
3030                               Philip Kerr   
3031                              Jodi Picoult   
3032                       Therese Anne Fowler   

                                            description            publisher  \
0     Aliens have taken control of the minds and bod...        Little, Brown   
1     A woman's happy marriage is shaken when she en...         St. Martin's   
2     A Massachusetts state investigator and his tea...               Putnam   
3     An aging porn queens aims to cap her career by...            Doubleday   
4                      A woman finds an unexpected love        Little, Brown   
...                                                 ...                  ...   
3028  Jake Fisher discovers that neither the woman h...               Dutton   
3029  Six friends meet in the 1970s at a summer arts...            Riverhead   
3030  Bernie Gunther, the Berlin cop, is sent to Smo...   Marian Wood/Putnam   
3031  A New Hampshire baker finds herself in the mid...  Emily Bestler/Atria   
3032  A novel based on the lives of Zelda and F. Sco...         St. Martin's   

                         title                       oid  \
0                     THE HOST  5b4aa4ead3089013507db18c   
1     LOVE THE ONE YOU'RE WITH  5b4aa4ead3089013507db18d   
2                    THE FRONT  5b4aa4ead3089013507db18e   
3                        SNUFF  5b4aa4ead3089013507db18f   
4         SUNDAYS AT TIFFANY’S  5b4aa4ead3089013507db190   
...                        ...                       ...   
3028                 SIX YEARS  5b4aa4ead3089013507dc593   
3029          THE INTERESTINGS  5b4aa4ead3089013507dc595   
3030      A MAN WITHOUT BREATH  5b4aa4ead3089013507dc597   
3031           THE STORYTELLER  5b4aa4ead3089013507dc598   
3032                         Z  5b4aa4ead3089013507dc59a   

      bestsellers_date.numberLong  published_date.numberLong  rank.numberInt  \
0                   1211587200000              1212883200000               2   
1                   1211587200000              1212883200000               3   
2                   1211587200000              1212883200000               4   
3                   1211587200000              1212883200000               5   
4                   1211587200000              1212883200000               6   
...                           ...                        ...             ...   
3028                1366416000000              1367712000000               9   
3029                1366416000000              1367712000000              11   
3030                1366416000000              1367712000000              13   
3031                1366416000000              1367712000000              14   
3032                1366416000000              1367712000000              16   

      rank_last_week.numberInt  weeks_on_list.numberInt  price.numberDouble  
0     

In [22]:
df.shape

(3033, 12)

In [23]:
df.columns

Index(['amazon_product_url', 'author', 'description', 'publisher', 'title',
       'oid', 'bestsellers_date.numberLong', 'published_date.numberLong',
       'rank.numberInt', 'rank_last_week.numberInt', 'weeks_on_list.numberInt',
       'price.numberDouble'],
      dtype='object')

In [24]:
df.dtypes

amazon_product_url              object
author                          object
description                     object
publisher                       object
title                           object
oid                             object
bestsellers_date.numberLong      int64
published_date.numberLong        int64
rank.numberInt                   int64
rank_last_week.numberInt         int64
weeks_on_list.numberInt          int64
price.numberDouble             float64
dtype: object

Ahora, usando todas las herramientas que hemos aprendido en esta sesión (indexación de filas y columnas, shape, dtypes, head, tail, columns, info, etc) explora tu dataset y responde a las siguientes preguntas:

    ¿Qué podemos saber de este dataset con tan sólo leer el nombre del archivo y los nombres de las columnas?
    R. Que se trata de una base de datos sobre los libros más vendidos.

    ¿Cuál es el tamaño (forma) de nuestro DataFrame? ¿Podríamos considerarlo un dataset grande o pequeño?
    R. El dataset tiene 3033 filas por 12 columnas, puede considerarse pequeño.

    ¿Crees que los nombres de las columnas son suficientemente descriptivos? ¿Podrían ser más claros o limpios?
    R. Creo que todos los nombres son bastante descriptivos, si acaso la columna oid podría resultar ambigua para quien no sabe del tema, pero revisando esta bien.

    ¿Qué tipos de datos tenemos?
    R. Objetos (posiblemente strings), enteros para las fechas, rankings, semanas en la lista y un flotante para el precio.

    ¿Qué significa el formato en el que tenemos las fechas?
    R. Que se trata de un timestamp

¿Qué tipo de preguntas podríamos responder usando los datos numéricos de este dataset?

- Cuáles son los libros más barato y caro de la lista y si eso influye en su volumen de venta
- Cuál es el libro que más tiempo ha permanecido en la lista
- El ranking de un libro va en aumento o descenso
- Desde su fecha de publicación, cuál es el libro que ha alcanzado estar en la lista de los más vendidos en el menor tiempo

¿Qué tipo de preguntas podríamos responder usando los datos no-numéricos?
    
- Que autor tiene más obras dentro de la lista de los más vendidos
- Una descripción de la obra
- Su casa editorial
- La liga para comprarlo